tests/components/notify/test_legacy.py

```diff
@@ -81,6 +112,394 @@ async def async_get_service(
+    assert get_service_called.call_args[0][1] == {}
+    get_service_called.reset_mock()
+
+    # Perform a reload
+    new_yaml_config_file = tmp_path / "configuration.yaml"
+    new_yaml_config = yaml.dump({"notify": [{"platform": "testnotify"}]})
+    new_yaml_config_file.write_text(new_yaml_config)
+
+    with patch.object(hass_config, "YAML_CONFIG_FILE", new_yaml_config_file):
+        await hass.services.async_call(
+            "testnotify",
+            SERVICE_RELOAD,
+            {},
+            blocking=True,
+        )
+        await hass.services.async_call(
+            "testnotify2",
+            SERVICE_RELOAD,
+            {},
+            blocking=True,
+        )
+        await hass.async_block_till_done()
+
+    # Check if the notify services from setup still exist
+    assert hass.services.has_service(notify.DOMAIN, "testnotify_a")
+    assert hass.services.has_service(notify.DOMAIN, "testnotify_b")
+    assert get_service_called.call_count == 1
+    assert get_service_called.call_args[0][0] == {"platform": "testnotify"}
+    assert get_service_called.call_args[0][1] is None
+
+    # Check if the dynamically notify services from setup were removed
+    assert not hass.services.has_service(notify.DOMAIN, "testnotify2_c")
+    assert not hass.services.has_service(notify.DOMAIN, "testnotify2_d")
+
+
+async def test_setup_platform_before_notify_setup(
+    hass: HomeAssistant, caplog: pytest.LogCaptureFixture, tmp_path: Path
+) -> None:
+    """Test trying to setup a platform before legacy notify service is setup."""
+    get_service_called = Mock()
+
+    async def async_get_service(hass, config, discovery_info=None):
+        """Get notify service for mocked platform."""
+        get_service_called(config, discovery_info)
+        targetlist = {"a": 1, "b": 2}
+        return NotificationService(hass, targetlist, "testnotify")
+
+    async def async_get_service2(hass, config, discovery_info=None):
+        """Get notify service for mocked platform."""
+        get_service_called(config, discovery_info)
+        targetlist = {"c": 3, "d": 4}
+        return NotificationService(hass, targetlist, "testnotify2")
+
+    # Mock first platform
+    mock_notify_platform(
+        hass, tmp_path, "testnotify", async_get_service=async_get_service
+    )
+
+    # Initialize a second platform testnotify2
+    mock_notify_platform(
+        hass, tmp_path, "testnotify2", async_get_service=async_get_service2
+    )
+
+    hass_config = {"notify": [{"platform": "testnotify"}]}
+
+    # Setup the second testnotify2 platform from discovery
+    load_coro = async_load_platform(
+        hass, Platform.NOTIFY, "testnotify2", {}, hass_config=hass_config
+    )
+
+    # Setup the testnotify platform
+    setup_coro = async_setup_component(hass, "notify", hass_config)
+
+    load_task = asyncio.create_task(load_coro)
+    setup_task = asyncio.create_task(setup_coro)
+
+    await asyncio.gather(load_task, setup_task)
```

Comment:
Side note: I don't think these two tests that say they are testing order of setup are doing what they say they are doing. There's no guarantee that one task completes before the other when using gather.